File: SmEngine/Libraries/MsHelper.vb
Imports System.Text.RegularExpressions
Imports Furcadia.Net.DreamInfo
Imports Monkeyspeak

Namespace Engine.Libraries

    Public NotInheritable Class MsLibHelper

        ''' <summary>
        ''' the last text the bot has seen, Usually the Triggering furre's message
        ''' </summary>
        Public Const MessageVariable As String = "%MESSAGE"

        ''' <summary>
        ''' the triggering furres shortname
        ''' </summary>
        Public Const ShortNameVariable = "%SHORTNAME"

        ''' <summary>
        ''' the triggering furres full name
        ''' </summary>
        Public Const TriggeringFurreNameVariable = "%NAME"

        ''' <summary>
        ''' the Uploaders name
        ''' </summary>
        Public Const DreamOwnerVariable = "%DREAMOWNER"

        ''' <summary>
        ''' the name of the dream we're in
        ''' </summary>
        Public Const DreamNameVariable = "%DREAMNAME"

        Public Const SmRegExOptions As RegexOptions = RegexOptions.Compiled

        ''' <summary>
        ''' Name of the connected furre, IE: the Bots name
        ''' </summary>
        Public Const BotNameVariable As String = "%BOTNAME"

        ''' <summary>
        ''' designated furre the bot recognizes as its controller
        ''' </summary>
        Public Const BotControllerVariable As String = "%BOTCONTROLLER"

        ''' <summary>
        ''' name of the furre getting banished ot unbanished
        ''' </summary>
        Public Const BanishNameVariable As String = "%BANISHNAME"

        ''' <summary>
        ''' comma seperated list of the furres on the banisl list
        ''' </summary>
        Public Const BanishListVariable As String = "%BANISHLIST"

        ''' <summary>
        ''' update Bot Constant Variables for the Current Triggering  Furre
        ''' <para/>
        ''' <see cref="TriggeringFurreNameVariable"/>
        ''' <para/>
        ''' <see cref="ShortNameVariable"/>
        ''' <para/>
        ''' <see cref="MessageVariable"/>
        ''' </summary>
        ''' <param name="ActivePlayer"></param>
        ''' <param name="MonkeySpeakPage"></param>
        Public Shared Sub UpdateTriggerigFurreVariables(ByRef ActivePlayer As IFurre, ByRef MonkeySpeakPage As Page)
            DirectCast(MonkeySpeakPage.GetVariable(MessageVariable), ConstantVariable).SetValue(ActivePlayer.Message)
            DirectCast(MonkeySpeakPage.GetVariable(ShortNameVariable), ConstantVariable).SetValue(ActivePlayer.ShortName)
            DirectCast(MonkeySpeakPage.GetVariable(TriggeringFurreNameVariable), ConstantVariable).SetValue(ActivePlayer.Name)
        End Sub

        ''' <summary>
        ''' update Bot Constant Variables for the Current Dream
        ''' <para/>
        ''' <see cref="DreamOwnerVariable"/>
        ''' <para/>
        ''' <see cref="DreamNameVariable"/>
        ''' </summary>
        ''' <param name="ActiveDream"></param>
        ''' <param name="MonkeySpeakPage"></param>
        Public Shared Sub UpdateCurrentDreamVariables(ByRef ActiveDream As Dream, ByRef MonkeySpeakPage As Page)
            DirectCast(MonkeySpeakPage.GetVariable(DreamOwnerVariable), ConstantVariable).SetValue(ActiveDream.Owner)
            DirectCast(MonkeySpeakPage.GetVariable(DreamNameVariable), ConstantVariable).SetValue(ActiveDream.Name)
        End Sub

        ''' <summary>
        ''' Reads a Double or a MonkeySpeak Variable
        ''' </summary>
        ''' <param name="reader">
        ''' <see cref="TriggerReader"/>
        ''' </param>
        ''' <param name="addIfNotExist">
        ''' Add Variable to Variable Scope is it Does not exist,
        ''' <para>
        ''' Default Value is False
        ''' </para>
        ''' </param>
        ''' <returns>
        ''' <see cref="Double"/>
        ''' </returns>
        Public Shared Function ReadVariableOrNumber(ByVal reader As TriggerReader,
                                         Optional addIfNotExist As Boolean = False) As Double
            Dim result = 0
            If reader.PeekVariable Then
                Dim value = reader.ReadVariable(addIfNotExist).Value.ToString
                Double.TryParse(value, result)
            ElseIf reader.PeekNumber Then
                result = reader.ReadNumber
            End If
            Return result
        End Function

    End Class

End Namespace

File: SmEngine/Libraries/MsHelper.cs
using Monkeyspeak;
using Furcadia.Net.DreamInfo;
using System.Text.RegularExpressions;

namespace Libraries
{
    public sealed class MsLibHelper
    {
        #region Public Fields

        /// <summary>
        /// The banish list variable
        /// </summary>
        public const string BanishListVariable = "%BANISHLIST";

        /// <summary>
        /// The banish name variable
        /// </summary>
        public const string BanishNameVariable = "%BANISHNAME";

        /// <summary>
        /// The bot controller variable
        /// </summary>
        public const string BotControllerVariable = "%BOTCONTROLLER";

        /// <summary>
        /// The bot name variable
        /// <para />
        /// %BOTNAME
        /// </summary>
        public const string BotNameVariable = "%BOTNAME";

        /// <summary>
        /// The dream name variable
        /// </summary>
        public const string DreamNameVariable = "%DREAMNAME";

        /// <summary>
        /// The dream owner variable
        /// </summary>
        public const string DreamOwnerVariable = "%DREAMOWNER";

        /// <summary>
        /// The triggering furre message variable
        /// </summary>
        public const string MessageVariable = "%MESSAGE";

        /// <summary>
        /// The sm reg ex options
        /// </summary>
        public const RegexOptions SmRegExOptions = RegexOptions.Compiled;

        /// <summary>
        /// The triggering furre name variable
        /// </summary>
        public const string TriggeringFurreNameVariable = "%NAME";

        /// <summary>
        /// The triggering furre short name variable
        /// </summary>
        public const string TriggeringFurreShortNameVariable = "%SHORTNAME";

        #endregion Public Fields

        #region Public Methods

        /// <summary>
        /// Reads the variable or number.
        /// <para />
        /// Default Value is False
        /// </summary>
        /// <param name="reader">The  <see cref="TriggerReader"/></param>
        /// <param name="addIfNotExist">Add Variable to Variable Scope is it Does not exist,</param>
        /// <returns><see cref="Double"/></returns>
        public static double ReadVariableOrNumber(TriggerReader reader, bool addIfNotExist)
        {
            double result = 0;
            if (reader.PeekVariable())
            {
                var value = reader.ReadVariable(addIfNotExist).Value.ToString();
                double.TryParse(value, out result);
            }
            else if (reader.PeekNumber())
            {
                result = reader.ReadNumber();
            }

            return result;
        }

        /// <summary>
        /// Updates the current dream variables.
        /// <para/>
        /// <see cref="DreamOwnerVariable"/>
        /// <para/>
        /// <see cref="DreamNameVariable"/>
        /// </summary>
        /// <param name="ActiveDream">The active dream.</param>
        /// <param name="MonkeySpeakPage">The monkey speak page.</param>
        public static void UpdateCurrentDreamVariables(Dream ActiveDream, Page MonkeySpeakPage)
        {
            if (!MonkeySpeakPage.HasVariable(DreamOwnerVariable))
            {
                MonkeySpeakPage.SetVariable(new ConstantVariable(DreamOwnerVariable, ActiveDream.Owner));
            }

            ((ConstantVariable)MonkeySpeakPage.GetVariable(DreamOwnerVariable)).SetValue(ActiveDream.Owner);

            if (!MonkeySpeakPage.HasVariable(DreamNameVariable))
            {
                MonkeySpeakPage.SetVariable(new ConstantVariable(DreamNameVariable, ActiveDream.Name));
            }

            ((ConstantVariable)MonkeySpeakPage.GetVariable(DreamNameVariable)).SetValue(ActiveDream.Name);
        }

        /// <summary>
        /// Updates the triggerig furre variables.
        /// </summary>
        /// <param name="ActivePlayer">The active player.</param>
        /// <param name="MonkeySpeakPage">The monkey speak page.</param>
        public static void UpdateTriggerigFurreVariables(Furre ActivePlayer, Page MonkeySpeakPage)
        {
            if (!MonkeySpeakPage.HasVariable(MessageVariable))
            {
                MonkeySpeakPage.SetVariable(new ConstantVariable(MessageVariable, ActivePlayer.Message));
            }

             ((ConstantVariable)MonkeySpeakPage.GetVariable(MessageVariable)).SetValue(ActivePlayer.Message);

            if (!MonkeySpeakPage.HasVariable(TriggeringFurreShortNameVariable))
            {
                MonkeySpeakPage.SetVariable(new ConstantVariable(TriggeringFurreShortNameVariable, ActivePlayer.ShortName));
            }

           ((ConstantVariable)MonkeySpeakPage.GetVariable(TriggeringFurreShortNameVariable)).SetValue(ActivePlayer.ShortName);

            if (!MonkeySpeakPage.HasVariable(TriggeringFurreNameVariable))
            {
                MonkeySpeakPage.SetVariable(new ConstantVariable(TriggeringFurreNameVariable, ActivePlayer.Name));
            }

           ((ConstantVariable)MonkeySpeakPage.GetVariable(TriggeringFurreNameVariable)).SetValue(ActivePlayer.Name);
        }

        #endregion Public Methods
    }
}